File: ruby/lib/br_validator/ie.rb
# frozen_string_literal: true

require_relative "ie/ac"
require_relative "ie/al"
require_relative "ie/am"
require_relative "ie/ap"
require_relative "ie/ba"
require_relative "ie/ce"
require_relative "ie/df"
require_relative "ie/es"
require_relative "ie/go"
require_relative "ie/ma"
require_relative "ie/mg"
require_relative "ie/ms"
require_relative "ie/mt"
require_relative "ie/pa"
require_relative "ie/pb"
require_relative "ie/pe"
require_relative "ie/pi"
require_relative "ie/pr"
require_relative "ie/rj"
require_relative "ie/rn"
require_relative "ie/ro"
require_relative "ie/rr"
require_relative "ie/rs"
require_relative "ie/sc"
require_relative "ie/se"
require_relative "ie/sp"
require_relative "ie/to"

module BrValidator
  # Validates, normalizes, and formats Brazilian Inscrição Estadual
  # numbers.
  #
  # Inscrição Estadual has no single national rule: each state (SEFAZ)
  # defines its own digit count and check-digit algorithm, so every UF is
  # modeled as its own module under BrValidator::Ie and registered in the
  # STATES map below. A UF absent from that map is simply not supported
  # yet.
  module Ie
    module_function

    STATES = {
      "AC" => Ac,
      "AL" => Al,
      "AM" => Am,
      "AP" => Ap,
      "BA" => Ba,
      "CE" => Ce,
      "DF" => Df,
      "ES" => Es,
      "GO" => Go,
      "MA" => Ma,
      "MG" => Mg,
      "MS" => Ms,
      "MT" => Mt,
      "PA" => Pa,
      "PB" => Pb,
      "PE" => Pe,
      "PI" => Pi,
      "PR" => Pr,
      "RJ" => Rj,
      "RN" => Rn,
      "RO" => Ro,
      "RR" => Rr,
      "RS" => Rs,
      "SC" => Sc,
      "SE" => Se,
      "SP" => Sp,
      "TO" => To
    }.freeze
    private_constant :STATES

    def resolve_state(uf)
      STATES[uf.strip.upcase]
    end
    private_class_method :resolve_state

    # Returns true if value is a valid Inscrição Estadual for uf.
    def is_valid?(value, uf)
      state = resolve_state(uf)

      return false if state.nil?

      state.is_valid?(value)
    end

    # Strips formatting and returns the canonical representation of value
    # for uf.
    #
    # If uf is not supported, value is returned unchanged.
    def normalize(value, uf)
      state = resolve_state(uf)

      return value if state.nil?

      state.normalize(value)
    end

    # Returns value in uf's standard human-readable representation.
    #
    # If uf is not supported or the normalized value has an invalid
    # length, value is returned unchanged.
    def format(value, uf)
      state = resolve_state(uf)

      return value if state.nil?

      state.format(value)
    end
  end
end

File: php/src/Ie.php
<?php

declare(strict_types=1);

namespace BrValidator;

use BrValidator\Ie\Ac;
use BrValidator\Ie\Al;
use BrValidator\Ie\Am;
use BrValidator\Ie\Ap;
use BrValidator\Ie\Ba;
use BrValidator\Ie\Ce;
use BrValidator\Ie\Df;
use BrValidator\Ie\Es;
use BrValidator\Ie\Go;
use BrValidator\Ie\Ma;
use BrValidator\Ie\Mg;
use BrValidator\Ie\Ms;
use BrValidator\Ie\Mt;
use BrValidator\Ie\Pa;
use BrValidator\Ie\Pb;
use BrValidator\Ie\Pe;
use BrValidator\Ie\Pi;
use BrValidator\Ie\Pr;
use BrValidator\Ie\Rj;
use BrValidator\Ie\Rn;
use BrValidator\Ie\Ro;
use BrValidator\Ie\Rr;
use BrValidator\Ie\Rs;
use BrValidator\Ie\Sc;
use BrValidator\Ie\Se;
use BrValidator\Ie\Sp;
use BrValidator\Ie\To;

/**
 * Validates, normalizes, and formats Brazilian Inscricao Estadual numbers.
 *
 * Inscricao Estadual has no single national rule: each state (SEFAZ)
 * defines its own digit count and check-digit algorithm, so every UF is
 * modeled as its own class under BrValidator\Ie and registered in the
 * STATES map below. A UF absent from that map is simply not supported
 * yet.
 */
final class Ie
{
    /** @var array<string, class-string> */
    private const STATES = [
        'AC' => Ac::class,
        'AL' => Al::class,
        'AM' => Am::class,
        'AP' => Ap::class,
        'BA' => Ba::class,
        'CE' => Ce::class,
        'DF' => Df::class,
        'ES' => Es::class,
        'GO' => Go::class,
        'MA' => Ma::class,
        'MG' => Mg::class,
        'MS' => Ms::class,
        'MT' => Mt::class,
        'PA' => Pa::class,
        'PB' => Pb::class,
        'PE' => Pe::class,
        'PI' => Pi::class,
        'PR' => Pr::class,
        'RJ' => Rj::class,
        'RN' => Rn::class,
        'RO' => Ro::class,
        'RR' => Rr::class,
        'RS' => Rs::class,
        'SC' => Sc::class,
        'SE' => Se::class,
        'SP' => Sp::class,
        'TO' => To::class,
    ];

    private function __construct()
    {
    }

    /** @return class-string|null */
    private static function resolveClass(string $uf): ?string
    {
        $key = strtoupper(trim($uf));

        return self::STATES[$key] ?? null;
    }

    /**
     * Reports whether $value is a valid Inscricao Estadual for $uf.
     */
    public static function isValid(string $value, string $uf): bool
    {
        $class = self::resolveClass($uf);

        if ($class === null) {
            return false;
        }

        return $class::isValid($value);
    }

    /**
     * Strips formatting and returns the canonical representation of
     * $value for $uf. If $uf is not supported, $value is returned
     * unchanged.
     */
    public static function normalize(string $value, string $uf): string
    {
        $class = self::resolveClass($uf);

        if ($class === null) {
            return $value;
        }

        return $class::normalize($value);
    }

    /**
     * Returns $value in $uf's standard human-readable representation. If
     * $uf is not supported or the normalized value has an invalid length,
     * $value is returned unchanged.
     */
    public static function format(string $value, string $uf): string
    {
        $class = self::resolveClass($uf);

        if ($class === null) {
            return $value;
        }

        return $class::format($value);
    }
}
